dev/doc/case-repr.md

```diff
@@ -0,0 +1,115 @@
+## Case representation
+
+Starting from Coq 8.14, the term representation of pattern-matching uses a so-called *compact form*. Compared to the previous representation, the major difference is that all type and term annotations on lambda and let abstractions that were present in branches and return clause of pattern-matchings were removed. In order to keep the ability to construct the old expanded form out of the new compact form, the case node also makes explicit data that was stealthily present in the expanded return clause, namely universe instances and parameters of the inductive type being eliminated.
```

Comment:
How about breaking up this very long line into 6 lines?

Can you be more specific than "stealthily present"?  Where was it present and how was it hidden?

Reply:
> How about breaking up this very long line into 6 lines?

You mean 6 paragraphs? The line layout is controlled by the renderer.

> Can you be more specific than "stealthily present"? Where was it present and how was it hidden?

The code relied on having this data around, but I only realized that by designing the new representation. So it would qualify as a "soft" invariant that could sometimes be repaired on the fly by doing retyping magic with enough finger crossing. By contrast the new representation turns this into a hard, explicit invariant.

Reply:
> You mean 6 paragraphs?

No.  The input line is several hundred characters long.  Please split it.  Look at the line numbers:

![image](https://user-images.githubusercontent.com/1253341/103449483-b2c73000-4c5d-11eb-985d-35f21d0050f7.png)